package/kedro_viz/server.py

```diff
@@ -180,6 +181,92 @@ def _get_pipeline_catalog_from_kedro14(env):
+        * While traversing, build up a dictionary of {node_id -> layers} for the node
+        that has already been visited.
+        * Turn the final {node_id -> layers} into a {layer -> layers} to represent the layers'
+        dependencies. Note: the key is a layer and the values are the parents of that layer,
+        just because that's the format toposort requires.
+        * Feed this layers dictionary to ``toposort`` and return the sorted values.
+        * Raise CircularDependencyError if the layers cannot be sorted topologically,
+        i.e. there are cycles among the layers.
+
+    Args:
+        nodes: A dictionary of {node_id -> node} represents the nodes in the graph.
+            A node's schema is:
+                {
+                    "type": str,
+                    "id": str,
+                    "name": str,
+                    "layer": Optional[str]
+                    ...
+                }
+        node_dependencies: A dictionary of {node_id -> set(child_node_ids)}
+            represents the direct dependencies between nodes in the graph.
+
+    Returns:
+        The list of layers sorted based on topological order.
+
+    Raises:
+        CircularDependencyError: When the layers have cyclic dependencies.
+    """
+    node_layers = {}  # map node_id to the layers that depend on it
+
+    def find_dependent_layers(node_id: str) -> Set[str]:
+        """For the given node_id, find all layers that depend on it in a depth-first manner.
+        Build up the node_layers dependency dictionary while traversing so each node is visited
+        only once.
+        """
+        if node_id in node_layers:
+            return node_layers[node_id]
+
+        node_layers[node_id] = set()
+
+        # for each dependent node of the given node_id,
+        # mark its layer and all layers that depend on it as dependent layers of the given node_id.
+        for dependent_node_id in node_dependencies[node_id]:
+            dependent_node = nodes[dependent_node_id]
+            dependent_layer = dependent_node.get("layer")
+            if dependent_layer is not None:
+                node_layers[node_id].add(dependent_layer)
+            node_layers[node_id].update(find_dependent_layers(dependent_node_id))
+
+        return node_layers[node_id]
+
+    # populate node_layers dependencies
+    for node_id in nodes:
+        find_dependent_layers(node_id)
+
+    # compute the layer dependencies dictionary based on the node_layers dependencies,
+    # represented as {layer -> set(parent_layers)}
+    layer_dependencies = defaultdict(set)
+    for node_id, dependent_layers in node_layers.items():
+        node_layer = nodes[node_id].get("layer")
+
+        # add the node's layer as a parent layer for all dependent layers
+        if node_layer is not None:
+            for layer in dependent_layers:
+                layer_dependencies[layer].add(node_layer)
```

Comment:
We should only add if `layer != node_layer`. If I read everything correctly, we haven't checked for this earlier, right?

Reply:
So interestingly enough, I thought of this yesterday. If you mark a node as a dependency of itself, `toposort` still works, so I omitted it to reduce mental tax. Happy to put it back in though.

```
>>> layers = {"raw": {"raw"}, "int": {"raw", "int"}}
>>> toposort_flatten(layers)
['raw', 'int']
```

Reply:
Oh, that's cool, didn't know about that 🙂 
Could you add a note please? So that whoever reads this knows that the case is covered.

Reply:
Done